include/device.h

```diff
@@ -208,6 +208,24 @@ extern "C" {
 
 struct device;
 
+/**
+ * @brief Device PM info
+ *
+ * @param usage device usage count
+ * @param enable device pm enable
+ * @param request device pm request
+ * @param status device pm request status
+ * @param dwork device pm work for asyncroneous request
+ * @param alert alert to notify the callers
+ */
+struct device_pm {
+	atomic_t usage;
+	atomic_t enable;
```

Comment:
A question on usage of `atomic` in Zephyr:  Of these four values the only one that appears to be used in a read-modify-write context is `usage`.  All others get and set single-word complete values (at least, based on a quick scan of the diff).  Is there a need for access to them to be atomic?

If yes, I'm concerned that a lot of other state values in Zephyr would have the same problem and there's major cleanup to do.

If no, use of atomic (e.g. with `enable`) is confusing, suggesting there's something tricky going on, and better to change them to int, bool, or enum values.

Reply:
My intention is to avoid the use of lock here and want to make sure we don't run into the case where one thread write and other one reads at the same time. what would you suggest?

Reply:
@andrewboie @andyross can you comment on the atomic side of things?

Reply:
The usage of atomics in the RMW contexts is obvious. Using atomics in non-RMW contexts is also valid in my opinion if saves us from using classic lock. Please look in the example below: 

```
struct foo {
    atomic_t enabled;
    (...)
}

void foo_enable(struct foo *foo)
{
  if (atomic_set(&foo->enabled, 1) == 0)
  {
      /* Protected code */
  }
  (...)
}

void foo_disable(struct foo *foo)
{
  if (atomic_set(&foo->enabled, 0) == 1)
  {
      /* Protected code */
  }
  (...)
}
```

The code above ensured that the `/* Protected code */` will be executed only once even if multiple threads would call simultaneously either `foo_enable()` or `foo_disable()`. In such situation we avoided classic lock with its penalties.

However such protection will fail if call to `foo_enable()` could be pre-empted by `foo_disable()` (or in reverse). In such situation usage if classic lock is unavoidable.

I do not have access to the discussed code (it was updated), but in general I would recommend using classic locks for non-RMW scenarios as a safer choice.